src/common/ad_mem.h

```diff
@@ -0,0 +1,327 @@
+	/** link chain on ad_blob LRU */
+	d_list_t		  ar_link;
+	/** arena type */
+	int			  ar_type;
+	/** refcount */
+	int			  ar_ref;
+	/** number of groups */
+	int			  ar_grp_nr;
+	/** unpublished arena */
+	bool			  ar_unpub;
+	/** being published */
+	bool			  ar_publishing;
+	/**
+	 * Arena is full, it's set to true when any type of group failed to allocate memory
+	 * and create more groups.
+	 * XXX: this is not enough, we should save failed allocatoin counter in matrics and
+	 * set arenea as full only if it encounters multiple failures.
+	 */
+	bool			  ar_full;
+	/** pointers for size binary search, it is only used by the DRAM mirror*/
+	struct ad_group_df	**ar_size_sorter;
+	/** pointers for address binary search, it is only used by the DRAM mirror*/
+	struct ad_group_df	**ar_addr_sorter;
+	/** reserved bits for groups */
+	uint64_t		  ar_bmap_rsv[ARENA_GRP_BMSZ];
+	/** metrics */
+	struct ad_group_metrics   ar_grp_mtcs[ARENA_GRP_SPEC_MAX];
+};
+
+/** page for each arena */
+struct ad_page {
+	/* the page in-use */
+	char			*pa_rpg;
+	/* reserved for checkpoint */
+	char			*pa_cpg;
+};
+
+/** reserved for future use */
+struct ad_page_extern {
+	struct ad_page		 pe_page;
+	d_list_t		 pa_link;
+	struct umem_store	*pa_store;
+};
+
+/* register up to 31 arenas types (type=0 is predefined) */
+#define ARENA_SPEC_MAX		32
+
+#define BLOB_HDR_SIZE		(32 << 10)
+#define BLOB_MAGIC		0xbabecafe
+
+#define AD_MEM_VERSION		1
+
+/** root data structure of ad-hoc allocator */
+struct ad_blob_df {
+	/** magic number */
+	uint32_t		bd_magic;
+	/** version number */
+	uint32_t		bd_version;
+	/** alignment */
+	int32_t			bd_unused;
+	/** number of registered arena types */
+	int32_t			bd_asp_nr;
+	/** specifications of registered arena types */
+	struct ad_arena_spec	bd_asp[ARENA_SPEC_MAX];
+	/** loading incarnation */
+	uint64_t		bd_incarnation;
+	/** it is DRAM reference of blob */
+	uint64_t		bd_back_ptr;
+	/** start address in the SPDK blob */
+	uint64_t		bd_addr;
+	/** capacity managed by the allocator */
+	uint64_t		bd_size;
+	/** size of each arena, default size is 16MB */
+	uint64_t		bd_arena_size;
+	/** allocated bits */
+	uint64_t		bd_bmap[0];
+};
+
+#define DUMMY_BLOB	"dummy.blob"
+
+/** DRAM blob open handle */
+struct ad_blob {
+	/** external umem store */
+	struct umem_store	 bb_store;
+	/** reserved: pages being checkpointed */
+	d_list_t		 bb_pgs_ckpt;
+	/** reserved: pages for external arenas */
+	d_list_t		 bb_pgs_extern;
+	/** arenas being reserved (not published), pinned at here */
+	d_list_t		 bb_ars_rsv;
+	/** unused arena LRU */
+	d_list_t		 bb_ars_lru;
+	/** group being reserved (not published), pinned at here */
+	d_list_t		 bb_gps_rsv;
+	/** unused group LRU */
+	d_list_t		 bb_gps_lru;
+	int			 bb_ars_lru_size;
+	int			 bb_gps_lru_size;
+	/** file descriptor of MD file */
+	int			 bb_fd;
+	/** reference counter */
+	int			 bb_ref;
+	/** is dummy blob, for unit test */
+	bool			 bb_dummy;
+	/** opened blob */
+	bool			 bb_opened;
```

Comment:
[minor] this might declared as to save some memory:
unsigned int                                      bb_dump:1,
                                                          bb_opened:1,

Reply:
this is the superblock, so I think it doesn't matter? if it's a data structure which has many instances, then I would agree.

Reply:
sure